File: 최윤경/정수 삼각형.js
/*
[접근]

- 모든 경로를 탐색하면 중복 계산 발생 → DP
- 아래 두 숫자 중 큰 값을 선택 → 최대 경로
- 아래에서 위로 올라오며 각 위치의 최대값 갱신
*/

function solution(triangle) {
  for (let level = triangle.length - 2; level >= 0; level--) {
    for (let index = 0; index <= level; index++) {
      triangle[level][index] += Math.max(triangle[level + 1][index], triangle[level + 1][index + 1]);
    }
  }

  return triangle[0][0];
}

File: 정선웅/정수_삼각형.java
import java.util.*;

class Solution {

    public int getMaxDp(int[][]dp, int i, int j) {
        if(j==i) {
            return dp[i-1][j-1];
        } else if(j==0) {
            return dp[i-1][j];
        } else {
            return Math.max(dp[i-1][j-1], dp[i-1][j]);
        }
    }

    public int solution(int[][] triangle) {
        int[][] dp = new int[triangle.length][triangle.length];
        dp[0][0] = triangle[0][0];

        int col = 0;
        for(int i=1; i<triangle.length; i++) {
            for(int j=0; j<triangle[i].length; j++) {
                dp[i][j] = getMaxDp(dp, i,j)+ triangle[i][j];
            }
        }

        int result = 0;
        for(int i=0; i<dp[dp.length-1].length; i++){
            result = Math.max(result, dp[dp.length-1][i]);
        }

        return result;
    }
}